File: strategy-game-backend/src/main/kotlin/de/ruegnerlukas/strategygame/backend/common/models/ProductionQueueEntry.kt
package de.ruegnerlukas.strategygame.backend.common.models

sealed class ProductionQueueEntry(
    val entryId: String,
    val collectedResources: ResourceCollection
) {
    abstract fun getTotalRequiredResources(): ResourceCollection
}


class BuildingProductionQueueEntry(
    entryId: String,
    collectedResources: ResourceCollection,
    val buildingType: BuildingType,
) : ProductionQueueEntry(entryId, collectedResources) {

    override fun getTotalRequiredResources(): ResourceCollection {
        return ResourceCollection.basic(buildingType.templateData.constructionCost)
    }

}


class SettlerProductionQueueEntry(
    entryId: String,
    collectedResources: ResourceCollection
) : ProductionQueueEntry(entryId, collectedResources) {

    override fun getTotalRequiredResources(): ResourceCollection {
        return ResourceCollection.basic( // TODO: AMOUNT CONFIGURABLE
            ResourceType.FOOD.amount(10f),
            ResourceType.WOOD.amount(5f)
        )
    }

}

File: frontend/src/core/models/productionQueueEntry.ts
import {BuildingType} from "./buildingType";

export interface ProductionQueueEntry {
    type: "building" | "settler"
    entryId: string,
    progress: number
}

export interface BuildingProductionQueueEntry extends ProductionQueueEntry {
    type: "building"
    buildingType: BuildingType;
}

export interface SettlerProductionQueueEntry extends ProductionQueueEntry {
    type: "settler"
}